File: lib/arguments.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

# -- Imports -------------------------------------------------------------------

from argparse import ArgumentParser, SUPPRESS

# -- Version -------------------------------------------------------------------

try:
    with open("VERSION","r") as v:
        version = v.readline()
except FileNotFoundError:
    version = "Unknown"

# -- Argument Parser -----------------------------------------------------------

parser = ArgumentParser(
    description="Just another programming language...", add_help=False)

# -- Defaults ------------------------------------------------------------------

default = parser.add_argument_group("Opciones Elementales")
default.add_argument("-h", "--help", "--ayuda", action="help",
                default=SUPPRESS,
                help="Muestra este mensaje de ayuda y sale.")
default.add_argument("-v", "--version", action="version",
                version="%(prog)s {}".format(version),
                help="Muestra la versión del programa y sale.")

# -- Inputs --------------------------------------------------------------------

inputs = parser.add_argument_group("Métodos de Lectura")
inputs.add_argument("archivo", type=str, nargs="?", default=None,
                help="Archivo con el código fuente del programa.")
inputs.add_argument("-i", "--interactive", "--interactivo",
                action="store_true",
                help="Ejecuta el programa en modo interactivo.")

# -- Parser --------------------------------------------------------------------

argv = parser.parse_args()
argv.__parser__ = parser.parse_args

File: lib/arguments.cpp
#include "arguments.hpp"

Arguments parseArguments(int argc, char* argv[])
{
    int option;
    Arguments args;

    if(argc == 1) args.help = true;
    else while((option = getopt(argc,argv,"f:hiv")) != EOF)
    {
        switch(option)
        {
            case 'f': args.filename = optarg;   break;
            case 'i': args.interactive = true;  break;
            case 'h': args.help = true;         break;
            case 'v': args.version = true;      break;
            default: args.help = true;          break;
        }
    }

    return args;
}

void show_help(std::string program)
{
    std::string help =
R"(Uso: )" + program + R"( [-i|-f archivo.alf] [-h] [-v]
    
Opciones de ejecución:
  -f ARCHIVO      Ejecuta el programa leyendo un archivo.
  -i              Ejecuta el programa de forma interactiva (REPL).
Opciones estándar:
  -h              Muestra esta ayuda y sale.
  -v              Muestra la versión y sale.)";

    std::cout << help << std::endl;
}

